crates/bevy_input/src/gamepad.rs

```diff
@@ -222,73 +538,154 @@ pub enum GamepadButtonType {
     Other(u8),
 }
 
-/// A button of a [`Gamepad`].
+/// The [`GamepadButtons`] [`component`](Component) is a collection of [`GamepadButtonType`] and their state during the current frame.
 ///
-/// ## Usage
+/// The [`entity`](Entity) representing a gamepad and its [`minimal components`](MinimalGamepad) are automatically managed.
 ///
-/// It is used as the generic `T` value of an [`ButtonInput`] and [`Axis`] to create `bevy` resources. These
-/// resources store the data of the buttons of a gamepad and can be accessed inside of a system.
+/// # Usage
 ///
-/// ## Updating
+/// The only way to obtain a [`GamepadButtons`] is by [`query`](Query).
 ///
-/// The gamepad button resources are updated inside of the [`gamepad_button_event_system`].
-#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Reflect)]
-#[reflect(Debug, Hash, PartialEq)]
-#[cfg_attr(
-    feature = "serialize",
-    derive(serde::Serialize, serde::Deserialize),
-    reflect(Serialize, Deserialize)
-)]
-pub struct GamepadButton {
-    /// The gamepad on which the button is located on.
-    pub gamepad: Gamepad,
-    /// The type of the button.
-    pub button_type: GamepadButtonType,
+/// # Examples
+///
+/// ```
+/// # use bevy_input::gamepad::{Gamepad, GamepadButtons, GamepadButtonType};
+/// # use bevy_ecs::system::Query;
+/// #
+/// fn gamepad_button_input_system(gamepads: Query<(&Gamepad, &GamepadButtons)>) {
+///     for (gamepad, buttons) in gamepads.iter() {
+///         if buttons.just_pressed(GamepadButtonType::North) {
+///             println!("{} just pressed North", gamepad.id())
+///         }
+///     }
+/// }
+/// ```
+#[derive(Component, Debug)]
+pub struct GamepadButtons {
```

Comment:
I'm thinking we should merge the `GamepadButtons` and `GamepadAxes` components into `Gamepad`. I think users would be much happier just querying for a Gamepad and reading the properties on it. Imo the `gamepad_input` example illustrates how much nicer the combined approach is:

```rust
fn gamepad_system(gamepads: Query<&Gamepad>) {
    for gamepad in &gamepads {
        if gamepad.just_pressed(GamepadButtonType::South) {
            info!("{:?} just pressed South", gamepad.id());
        } else if gamepad.just_released(GamepadButtonType::South) {
            info!("{:?} just released South", gamepad.id());
        }

        let right_trigger = gamepad.button(GamepadButtonType::RightTrigger2).unwrap();
        if right_trigger.abs() > 0.01 {
            info!(
                "{:?} RightTrigger2 value is {}",
                gamepad.id(),
                right_trigger
            );
        }

        let left_stick_x = gamepad.get(GamepadAxisType::LeftStickX).unwrap();
        if left_stick_x.abs() > 0.01 {
            info!("{:?} LeftStickX value is {}", gamepad.id(), left_stick_x);
        }
    }
}
```

`GamepadSettings` can/should probably still be a separate component, as it feels more niche and introduces a solid chunk of additional complexity.

Reply:
Yeah I don't think performance is a major concern in this context. The number of gamepads will always be small and the work being done isn't particularly expensive.